Cluedb dfs solve

jpalpant opened the pull request:
Not actually a DFS-based solver anymore, but ClueDBSolver is now able to effectively find potential solutions using exact matches in the ClueDB it uses.  It gives each solution a confidence of 1.0, but the solutions are returned sorted from best to worst in terms of number of clues with answers.

Generally, ClueDBSolver finds a LOT of solutions (hundreds to thousands), and it doesn't have a good way to determine how good a solution is since as far as it was concerned, every word it played was an exact match for the clue it was given.  That could use some work.  But for a first step it's quick and smart and reasonably memory efficient, finding it's thousands of solutions in less than a second on most of my tests.

Take a look!  (You'll probably have to change solve_xword.py like I did.  I also recommend puzzle daily-2015-11-27 for a puzzle with only a few solutions, and daily-2016-01-19 for a puzzle with many)

Commit:
Change implementation of ClueDBSolver

A pure depth-first search of combinations of clue answers proved
infeasible, so, new approach

It is known that two different across clues cannot conflict with each
other - likewise, two down clues cannot conflict with each other.  So,
using set product, we can construct a set which contains all
combinations of answers for all clues that have answers in the ClueDB
for one direction.  A set of all possible ways to answer the across
clues is called an "across fill", while a set of all possible
combinations of answers to down clues is called a "down fill".  

Starting with each possible across fill, we will do a breadth first
search using each down fill - first attempting to use the entire down
fill, and then attempting to use the down fill but with one word
removed, and then two, etc. etc.

Because we starting with using one answer for every clue and then
attempted less optimal answers, we can terminate branches as soon as one
puzzle is valid.

Commit:
Implement apply_fill to apply a fill set to a crossword

Also implement first (untested) breadth-first search starting from a
complete across fill

Commit:
Implement set-based search for solutions

The new set-based search determines conflicts in advance and uses known
conflicts to quickly determine the largest nonconflicting subset of clue
answers from the ClueDB

solve() does not return the correct type net, will finish that shortly

Commit:
Update documentation

Commit:
Fix return type of solve() - now returns a list of Crosswords

Solve follows the API set out in Solver - before it was returning other
things that I wanted to look at

Commit:
Format code and clean up docstrings

timpalpant commented on lib-python/littleboxes/xword.py:
```diff
@@ -1,22 +1,24 @@
 import logging
 import puz
 
+
```
What is the formatting styler you are using? I should probably set mine up to standardize formatting before committing.

jpalpant replied:
Ah I forgot about that!  It's reflex now - Eclipse has autopep8 built in as a code formatter, it does some nice indenting for you, keeps lines from being too long sometimes, and makes sure there aren't extra whitespace characters.  I can stop using it also, or make sure to undo when I accidently do, but it's pretty nice if you can find how to use it outside of eclipse/however you code
![screen shot 2016-02-27 at 4 10 52 pm](https://cloud.githubusercontent.com/assets/9629191/13375424/d0cd2df8-dd6c-11e5-8f0c-096bded5ab67.png)

timpalpant replied:
autopep8 is great! I've added it to my editor and will fix up the existing violations

timpalpant commented on lib-python/littleboxes/solver/cluedb_solver.py:
```diff
@@ -1,15 +1,225 @@
+        start_conflicts = time.time()
+        conflict_map = self.build_conflict_map(xword, possible_answers)
+
+        print 'Conflicts map build in {0} seconds'.format(
```
Can you make these logging statements?

jpalpant replied:
Yeah!  ClueDBSolver doesn't have a logger to call it's own, should I just use the root logger with the default formatting?  i.e. logging.info('stuff')

timpalpant replied:
I like the log4j convention of creating a logger for each class using the full class name (in this case `littleboxes.solver.ClueDBSolver`). It makes it easy to turn on/off logging from different parts of the application as needed. What do you think? (Using the root logger is another common choice).

jpalpant replied:
I didn't realize it was a convention, but that's exactly what I did for senior design
![screen shot 2016-02-27 at 6 17 20 pm](https://cloud.githubusercontent.com/assets/9629191/13376051/697273ae-dd7e-11e5-8e17-cc225ea64bda.png)

So I like it, I'll do that

jpalpant commented:
Slight problem with the solver - it's written to be exhaustive, so it takes every possible set of solutions to across clues and every possible set of solutions to down clues and takes the product of those two sets.  If the ClueDB doesn't have many ideas on what to do for the clues, that's fine - but sometimes, it seems, it has lots of potential words:

> INFO:root:Loading dictionary
> INFO:root:Solving puzzle
> INFO:littleboxes.solver.ClueDBSolver:Conflicts map build in 0.0759859085083 seconds
> INFO:littleboxes.solver.ClueDBSolver:Size of across fills: 13271040
> INFO:littleboxes.solver.ClueDBSolver:Size of down fills: 640

(Puzzle daily-1994-04-07)

This one would have had 8,493,465,600 sets of words to check for conflicts and then for uniqueness.  That's far too many, and who knows what other puzzles might have.  I might need to do something about the fact that the time and space complexity expands with the product of the number of solutions the ClueDB returns for each clue, but I don't have any ideas on how to reduce that yet, since that's just asking the ClueDB for perfect matches.

Commit:
Implement logging

Next up, remove some of the inefficient things, like loading entire sets
into memory when they could just as easily be iterators

timpalpant commented:
I'm not sure I fully understand the algorithm yet, but might it be missing solutions in which some across answers win conflicts and other down answers win conflicts? Imagine there are two disjoint parts of the puzzle (say top-left and bottom-right corners). Both corners have across/down conflicts, but you might have a solution in which the across answers are used in the top-left and the down answers used in the bottom-right.

Also, you might want solutions in which not all of the possible answers are played (in case one of them is incorrect).

This doesn't help the combinatorial explosion problem though :)

jpalpant commented:
That is definitely the case, it _would_ miss all the mixed-conflict resolution solutions wouldn't it.  Which increases the combinatorial explosion by another factor of something factorial, I bet.  I'll think about that while I'm thinking about updating for the new Crossword class and thinking about not loading all my iterators into memory (my three tasks for the time being).

timpalpant commented:
Here's an idea that may or may not be useful. The ClueDB gives us a set of possible answers for some clues in the crossword puzzle. (The Dictionary does too). Some of these answers are mutually exclusive (they indicate different letters for a single square), and we want to enumerate all valid (non-conflicting) solutions that result from placing a subset of the possible answers.

Label the set of possible answers returned from the ClueDB as A(1,1), A(1,2), D(2,1), etc., where A(1,1) indicates a possible answer for 1-across, A(1,2) indicates a second, different answer for 1-across, D(2,1) indicates a possible answer for 2-down, etc. Consider each possible answer a vertex in a graph, with edges between all pairs of vertices that are compatible (non-conflicting):

![graph](https://cloud.githubusercontent.com/assets/814226/13416696/aa88718a-df32-11e5-9cd1-852c3e8627ba.png)
1. All answers for the same clue conflict (e.g. A(1,1) and A(1,2)).
2. All answers for different across clues are connected (e.g. A(1,1) and A(2,1)). Similarly for down clues.
3. Some across answers will conflict with some down answers (e.g. A(1,1) and D(2,1)).

Valid, non-conflicting solutions are [cliques](https://en.wikipedia.org/wiki/Clique_%28graph_theory%29) of this graph: all of the selected answers must be compatible/connected to all other answers. We want to [enumerate all of the cliques](https://en.wikipedia.org/wiki/Clique_problem#Listing_all_maximal_cliques) of this graph. There are at most 3^(n/3) maximal cliques (n: number of vertices), which can be listed with the [Bron-Kerbosch algorithm](https://en.wikipedia.org/wiki/Bron–Kerbosch_algorithm) (Hopefully with the above restrictions the actual number is smaller). Then all subgraphs of the maximal cliques are also (smaller) cliques. (In the solver, we probably want to yield maximal cliques first and then smaller solutions).

jpalpant commented:
That makes sense and looks like a good thing to try!  Especially because it does extend so well to the DictionarySolver (which was something I was thinking about while doing ClueDB).  One good thing to think about is - do we really need to return non-maximal cliques?  It'll increase the size of the set of returned solutions by a huge factor.  If we're going to remove words from maximal solutions where all words are perfect matches in our database, we should do it selectively, like if removing a word is expected to open up a large branch of possible solutions or if a word is determined to be unlikely (both of which would be fun heuristics to write).

jpalpant commented:
It bothers me that the branch name doesn't reflect what is actually going on, so I deleted this branch and pull request to make a new branch name that is more general.  It closes the pull request, apparently, but the ClueDB branch wasn't near ready so I'll make a new one once cluedb_solver (the new branch) uses the new crossword class and I'll give a few tries at the graph method.
